File: client-android/lib/src/main/kotlin/stasis/client_android/lib/api/clients/DefaultServerBootstrapEndpointClient.kt
package stasis.client_android.lib.api.clients

import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okio.ByteString
import stasis.client_android.lib.api.clients.exceptions.EndpointFailure
import stasis.client_android.lib.api.clients.exceptions.InvalidBootstrapCodeFailure
import stasis.client_android.lib.api.clients.internal.ClientExtensions
import stasis.client_android.lib.model.server.devices.DeviceBootstrapParameters
import stasis.client_android.lib.security.HttpCredentials
import stasis.client_android.lib.security.HttpCredentials.Companion.withCredentials
import stasis.client_android.lib.utils.AsyncOps
import stasis.client_android.lib.utils.AsyncOps.asyncRetryWith
import stasis.client_android.lib.utils.NonFatal.nonFatal
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Failure
import stasis.client_android.lib.utils.Try.Success

class DefaultServerBootstrapEndpointClient(
    serverBootstrapUrl: String,
) : ServerBootstrapEndpointClient, ClientExtensions() {
    override val server: String = serverBootstrapUrl.trimEnd { it == '/' }

    override val credentials: suspend () -> HttpCredentials = { HttpCredentials.None }

    override val retryConfig: AsyncOps.RetryConfig = AsyncOps.RetryConfig.Default

    override suspend fun execute(bootstrapCode: String): Try<DeviceBootstrapParameters> {
        val request = Request.Builder()
            .url("$server/v1/devices/execute")
            .put(ByteString.EMPTY.toRequestBody())
            .withCredentials(HttpCredentials.OAuth2BearerToken(token = bootstrapCode))
            .build()

        return try {
            val response = client.newCall(request).asyncRetryWith(retryConfig)
            when {
                response.isSuccessful -> Success(response.toRequiredModel())
                response.code == StatusUnauthorized -> Failure(InvalidBootstrapCodeFailure())
                else -> Failure(
                    EndpointFailure(
                        "Server [$server] responded with [${response.code} - ${response.message}]"
                    )
                )
            }
        } catch (e: Throwable) {
            Failure(e.nonFatal())
        }
    }

    companion object {
        const val StatusUnauthorized: Int = 401
    }
}

File: client-ios/lib/Sources/StasisClientLib/Api/Clients/DefaultServerBootstrapEndpointClient.swift
import Foundation

public struct DefaultServerBootstrapEndpointClient: ServerBootstrapEndpointClient {
    public let server: String
    public let http: HttpClient

    public init(serverBootstrapUrl: String, retryConfig: RetryConfig = .default) {
        self.server = serverBootstrapUrl.trimmedTrailingSlash
        self.http = HttpClient(
            credentialsProvider: StaticHttpCredentialsProvider(.none),
            retryConfig: retryConfig
        )
    }

    init(serverBootstrapUrl: String, http: HttpClient) {
        self.server = serverBootstrapUrl.trimmedTrailingSlash
        self.http = http
    }

    public func execute(bootstrapCode: String) async throws -> DeviceBootstrapParameters {
        var request = URLRequest(url: URL(string: "\(server)/v1/devices/execute")!)
        request.httpMethod = "PUT"
        request.httpBody = Data()
        request.setCredentials(.oauth2BearerToken(token: bootstrapCode))

        let (data, response) = try await http.send(request)

        switch response.statusCode {
        case 200..<300:
            return try JSONCoders.decoder().decode(DeviceBootstrapParameters.self, from: data)
        case HTTPURLResponse.statusUnauthorized:
            throw InvalidBootstrapCodeFailure()
        default:
            let message = HTTPURLResponse.localizedString(forStatusCode: response.statusCode)
            throw EndpointFailure(
                message: "Server [\(server)] responded with [\(response.statusCode) - \(message)]"
            )
        }
    }
}
